pkg/controller/machinemanagement/machinemanagement_controller.go

```diff
@@ -82,6 +85,50 @@ func AddToManager(mgr manager.Manager, r reconcile.Reconciler, concurrentReconci
 		return err
 	}
 
+	err = mgr.GetFieldIndexer().IndexField(context.TODO(), &hivev1.ClusterDeployment{}, "spec.secrets.secretName", func(o client.Object) []string {
+		var res []string
+		cd := o.(*hivev1.ClusterDeployment)
+		if utils.CredentialsSecretName(cd) != "" {
+			res = append(res, utils.CredentialsSecretName(cd))
+		}
+		if cd.Spec.PullSecretRef != nil {
+			res = append(res, cd.Spec.PullSecretRef.Name)
+		}
+		if cd.Spec.Provisioning.SSHPrivateKeySecretRef != nil {
+			res = append(res, cd.Spec.Provisioning.SSHPrivateKeySecretRef.Name)
+		}
+		return res
+	})
+	if err != nil {
+		log.WithField("controller", ControllerName).WithError(err).Error("Error indexing cluster deployment secrets")
+	}
+
+	// Watch for changes to Secret referenced by cluster deployment
+	err = c.Watch(&source.Kind{Type: &corev1.Secret{}}, handler.EnqueueRequestsFromMapFunc(func(a client.Object) []reconcile.Request {
+		retval := []reconcile.Request{}
+
+		secret := a.(*corev1.Secret)
+		if secret == nil {
+			// Wasn't a Secret, bail out. This should not happen.
+			log.Errorf("Error converting MapObject.Object to Secret. Value: %+v", a)
+			return retval
+		}
```

Comment:
this will not prevent a panic when a is not type `*corev1.Secret` i think. we probably need `typed, ok := a.(*corev1.Secret)`